pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go

```diff
@@ -0,0 +1,400 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package assemble
+
+import (
+	runtimev1alpha1 "github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplane/crossplane-runtime/pkg/fieldpath"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog/v2"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha2"
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+	ctrlutil "github.com/oam-dev/kubevela/pkg/controller/utils"
+	"github.com/oam-dev/kubevela/pkg/oam"
+	"github.com/oam-dev/kubevela/pkg/oam/util"
+)
+
+// NewAppManifests create a AppManifests
+func NewAppManifests(appRevision *v1beta1.ApplicationRevision) *AppManifests {
+	return &AppManifests{AppRevision: appRevision}
+}
+
+// AppManifests contains configuration to assemble resources recorded in the ApplicationRevision.
+// 'Assemble' means expand Application(Component and Trait) into K8s resource and get them ready to go, to be emitted
+// into K8s
+type AppManifests struct {
+	AppRevision     *v1beta1.ApplicationRevision
+	WorkloadOptions []WorkloadOption
+
+	appComponents []applicationComponent
+	components    []*v1alpha2.Component
+
+	appName        string
+	appNamespace   string
+	appLabels      map[string]string
+	appAnnotations map[string]string
+	appOwnerRef    *metav1.OwnerReference
+
+	assembledWorkloads map[string]*unstructured.Unstructured
+	assembledTraits    map[string][]*unstructured.Unstructured
+	// key is workload reference, values are the references of scopes the workload belongs to
+	referencedScopes map[runtimev1alpha1.TypedReference][]runtimev1alpha1.TypedReference
+
+	finalized bool
+	err       error
+}
+
+// this is a helper struct to replace v1alpha2.ApplicationConfiguration for this moment
+type applicationComponent struct {
+	revisionName string
+	traits       []v1alpha2.ComponentTrait
+	scopes       []runtimev1alpha1.TypedReference
+}
+
+// WorkloadOption will be applied to each workloads AFTER it has been assembled by generic rules shown below:
+// 1) use component name as workload name
+// 2) use application namespace as workload namespace if unspecified
+// 3) set application as workload's owner
+// 4) pass all application's labels and annotations to workload's
+// Component and ComponentDefinition are enough for caller to manipulate workloads.
+// Caller can use below labels of workload to get more information:
+// - oam.LabelAppName
+// - oam.LabelAppRevision
+// - oam.LabelAppRevisionHash
+// - oam.LabelAppComponent
+// - oam.LabelAppComponentRevision
+type WorkloadOption interface {
+	ApplyToWorkload(workload *unstructured.Unstructured, comp *v1alpha2.Component, compDefinition *v1beta1.ComponentDefinition) error
```

Comment:
why `v1alpha2.Component` is still exist concept?

Reply:
for helm field in component, there's a WorkloadOption still need it.

Reply:
We should find a replace, we need to remove dependency on AppConfig/AppContext/Component